### PR TITLE
Attribute restrictions

### PR DESCRIPTION
**Pull request by SimonCockx**
Addresses #863 

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

### CODE REVIEW

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/object/ModelObjectBuilderGenerator.xtend [hunk 1]
```diff
@@ -59,259 +60,433 @@ class ModelObjectBuilderGenerator {
...
-				public «thisName» set«attribute.name.toFirstUpper»Value(«attribute.toJavaType» «scope.getIdentifierOrThrow(attribute)») {
-					this.getOrCreate«attribute.name.toFirstUpper»().setValue(«scope.getIdentifierOrThrow(attribute)»);
-					return this;
-				}
 			«ENDIF»
 		«ENDIF»
+		«IF prop.parentProperty !== null»«doCompatibleSetter(javaType, prop, prop.parentProperty, scope)»«ENDIF»
+		'''
+	}
+	private def StringConcatenationClient doCompatibleSetter(JavaPojoInterface javaType, JavaPojoProperty originalProp, JavaPojoProperty parentProp, JavaScope scope) {
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/object/ModelObjectBuilderGenerator.xtend:295]: TODO: clean this up - too much copy-pasting.

Comment (SimonCockx) [reply rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/object/ModelObjectBuilderGenerator.xtend:295]: Also add appropriate @Override annotations

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/object/ModelObjectGenerator.xtend [hunk 2]
```diff
@@ -234,36 +218,45 @@ class ModelObjectGenerator {
...
-			«FOR attribute : attributes»
-				«method(Optional, "ofNullable")»(get«attribute.name.toFirstUpper»()).ifPresent(builder::set«attribute.name.toFirstUpper»);
+			«IF extended»super.setBuilderFields(builder);«ENDIF»
+			«FOR prop : properties»
+				«method(Optional, "ofNullable")»(«prop.getterName»()).ifPresent(builder::set«prop.name.toFirstUpper»);
 			«ENDFOR»
 		}
 		'''
 	}
+	private def StringConcatenationClient derivedIncompatibleGettersForProperty(JavaExpression originalField, JavaPojoProperty prop, JavaScope scope) {
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/object/ModelObjectGenerator.xtend:228]: TODO: too much copy-pasting - clean up.

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/reports/TabulatorGenerator.xtend [hunk 3]
```diff
@@ -458,7 +459,7 @@ class TabulatorGenerator {
 		if (rType instanceof RDataType) {
 			val nestedTabulator = scope.getIdentifierOrThrow(rType.toNestedTabulatorInstance)
 			'''
-			«FieldValue» «resultId» = «Optional».ofNullable(«inputParam».get«attr.name.toFirstUpper»())
+			«FieldValue» «resultId» = «Optional».ofNullable(«inputParam».«javaType.findProperty(attr.name).getterName»())
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/reports/TabulatorGenerator.xtend:462]: TODO: refactor common code...

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/JavaPojoInterface.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+			}
+			currentProp = currentProp.getParentProperty();
+		}
+		// Fallback: no compatible type found
+		return prop;
+	}
+	public JavaPojoProperty findProperty(String propertyName) {
+		return getAllProperties().stream().filter(prop -> prop.getName().equals(propertyName))
+			.findAny()
+			.orElseThrow();
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/JavaPojoInterface.java:49]: Add useful error message

**Comment by SimonCockx** [comment]
Add auto-completion

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/RJavaPojoInterface.java [hunk 5]
```diff
@@ -16,42 +16,133 @@
...
+			}
+			return "get" + StringUtils.capitalize(propertyName) + "RestrictedAsSingle" + specializedType.getSimpleName();
+		} else {
+			// Type to other type
+			return "get" + StringUtils.capitalize(propertyName) + "RestrictedAs" + specializedType.getSimpleName();
+		}
 	}
 
 	@Override
 	public boolean isSubtypeOf(JavaType other) {
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/RJavaPojoInterface.java:144]: TODO: review this implementation

File: rosetta-ide/package-lock.json
Comment (SimonCockx) [inline rosetta-ide/package-lock.json:1]: Remove this file

File: rosetta-lang/model/Rosetta.xcore [hunk 6]
```diff
@@ -126,7 +122,7 @@ class RosettaTypeAlias extends RosettaRootElement, RosettaType, RosettaTyped, Ro
 }
 
 
-class RosettaMetaType extends RosettaRootElement, RosettaTypedFeature, RosettaType, RosettaSymbol {
```
Comment (SimonCockx) [inline rosetta-lang/model/Rosetta.xcore:129]: Also change `RosettaTypedFeature` into `RosettaFeature`!

**Review by davidalk — APPROVED** [review]

**Comment by hugohills-regnosys** [comment]
https://github.com/finos/common-domain-model/issues/3364

**Pull request merged**